url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/model/Model.java

```diff
@@ -0,0 +1,334 @@
+                            for (int p = 0; p < 3; ++p) {
+                                shape.rotate(rotation);
+                            }
+                            return false;
+                        }
+                    }
+
+                }
+            }
+        }
+
+        return true;
+    }
+
+    public boolean rotateYes(Shape shape) {
+        return rotate(shape, YES);
+    }
+
+    private boolean rotateReverse(Shape shape, int rotation) {
+        if (rotate(shape, rotation)) {
+            if (rotate(shape, rotation)) {
+                if (rotate(shape, rotation)) {
+                    return true;
+                } else {
+                    shape.rotate(rotation);
+                    shape.rotate(rotation);
+                }
+            } else {
+                shape.rotate(rotation);
+                shape.rotate(rotation);
+                shape.rotate(rotation);
+            }
+        }
+        return false;
+    }
+
+    public boolean rotateYesReverse(Shape shape) {
+        return rotateReverse(shape, YES);
+    }
+
+    public boolean rotateNo(Shape shape) {
+        return rotate(shape, NO);
+    }
+
+    public boolean rotateNoReverse(Shape shape) {
+        return rotateReverse(shape, NO);
+    }
+
+    public boolean rotateMaybe(Shape shape) {
+        return rotate(shape, MAYBE);
+    }
+
+    public boolean rotateMaybeReverse(Shape shape) {
+        return rotateReverse(shape, MAYBE);
+    }
+
+    public boolean incrementLevel(Shape shape) {
+        boolean[][][] shapeMatrix = shape.getMatrix();
+        boolean band = false;
+        boolean unir = false;
+        int maxZ = 0;
+        int x, y, z;
+        for (int i = 0; i < shapeMatrix.length && !unir; ++i) {
+            for (int j = 0; j < shapeMatrix[i].length && !unir; ++j) {
+                band = false;
+                maxZ = 0;
+                for (int k = 0; k < shapeMatrix[0][0].length; ++k) {
+                    if (shapeMatrix[i][j][k]) {
+                        band = true;
+                        maxZ = Math.max(maxZ, k);
+                    }
+                }
+                x = i + shape.getDesfX();
+                y = j + shape.getDesfY();
+                z = maxZ + shape.getLevel() + 1;
+                if (x < matrix.length && y < matrix[0].length && z < matrix[0][0].length) {
+                    if (band && matrix[x][y][z]) {
+                        unir = true;
+                    }
+                }
+            }
+        }
+        if (unir) {
+            for (int i = 0; i < shapeMatrix.length; ++i) {
+                for (int j = 0; j < shapeMatrix[i].length; ++j) {
+                    for (int k = 0; k < shapeMatrix[i][j].length; ++k) {
+                        if (shapeMatrix[i][j][k]) {
+                            matrix[i + shape.getDesfX()][j + shape.getDesfY()][k + shape.getLevel()] = true;
+                        }
+                    }
+                }
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 254:262, 300:308.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.